File: exercise/99/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void recoverTree(TreeNode* root) {
        TreeNode *first = NULL, *second = NULL;
        TreeNode *pre = NULL, *cur = root;
        while (cur) {
            if (!cur->left) {
                if (pre && cur->val < pre->val) {
                    if (first)
                        second = cur;
                    else {
                        first = pre;
                        second = cur;
                    }
                }
                pre = cur;
                cur = cur->right;
                continue;
            }
            pre = cur->left;
            while (pre->right && pre->right != cur) {
                pre = pre->right;
            }
            if (pre->right == cur) {
                pre->right = NULL;
                if (pre && cur->val < pre->val) {
                    if (first)
                        second = cur;
                    else {
                        first = pre;
                        second = cur;
                    }
                }
                pre = cur;
                cur = cur->right;
            } else {
                pre->right = cur;
                pre = cur;
                cur = cur->left;
            }
        }
        int t = first->val;
        first->val = second->val;
        second->val = t;
    }
};

File: exercise/99/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def update(self, root: TreeNode, last:TreeNode, first: TreeNode, second: TreeNode) -> bool:
        if last and root.val < last.val:
            if first:
                second = root
                return True
            else:
                first, second = last, root
        last = root
        return False

    def recoverTree(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        last = None
        first, second = None, None
        while root:
            if root.left is None:
                if self.update(root, last, first, second):
                    break
                root = root.right
                continue
            pre = root.left
            while pre.right and pre.right != root:
                pre = pre.right
            if pre.right:
                pre.right = None
                if self.update(root, last, first, second):
                    break
                root = root.right
            else:
                pre.right = root
                root = root.left
        first.val, second.val = second.val, first.val

